iree/turbine/kernel/wave/gather_to_shared.py

```diff
@@ -0,0 +1,364 @@
+    logger.info(f"total_number_of_elements={total_number_of_elements}")
+    expected_number_of_loads = ceildiv(total_number_of_elements, max_elements_per_store)
+    logger.info(f"expected_number_of_loads={expected_number_of_loads}")
+    elements_per_thread = ceildiv(
+        total_number_of_elements, expected_number_of_loads * total_number_of_threads
+    )
+    logger.info(f"elements_per_thread={elements_per_thread}")
+
+    vector_width = elements_per_thread * bitwidth
+    load_width = get_load_width(supported_load_widths, vector_width)
+    if load_width is None:
+        logger.error(f"No supported load width found for bitwidth={vector_width}")
+        return None
+
+    logger.info(f"load_width={load_width}")
+
+    # Get supported load width for the given bitwidth and if they are not
+    # equal then we need to adjust the number of loads and elements per thread.
+    ratio = vector_width // load_width
+    logger.info(f"ratio={ratio}")
+    expected_number_of_loads *= ratio
+    elements_per_thread //= ratio
+
+    if materialized_shape[-1] % elements_per_thread != 0:
+        logger.info(
+            f"materialized_shape[-1]={materialized_shape[-1]} is not divisible by "
+            f"elements_per_thread={elements_per_thread}"
+        )
+        return None
+
+    return GatherToSharedConfig(
+        materialized_shape,
+        elements_per_thread,
+        expected_number_of_loads,
+    )
+
+
+def create_writes(
+    read: Read,
+    write: Write,
+    materialized_shape: list[IndexSymbol],
+    elements_per_thread: int,
+    expected_number_of_loads: int,
+    total_number_of_threads: int,
+    thread_id: IndexExpr,
+    symbolic_shape: list[IndexSymbol],
+    wave_subs: dict[IndexSymbol, IndexExpr],
+    element_type: "DataType",
+) -> defaultdict[fx.Node, list[Write]]:
+    """
+    Create new writes for the given read and write.
+    """
+    elements_per_wave = elements_per_thread * total_number_of_threads
+    logger.info(f"elements_per_wave={elements_per_wave}")
+
+    # For index delinearization, assume our shape in `elements_per_thread` chunks.
+    materialized_shape_adjusted = list(materialized_shape)
+    materialized_shape_adjusted[-1] = materialized_shape[-1] // elements_per_thread
+
+    logger.info(f"materialized_shape_adjusted={materialized_shape_adjusted}")
+
+    # GatherToLDS writes `elements_per_wave` elements contiguously to LDS, so we
+    # cannot have any padding if it crosses a array row boundary.
+    drop_padding = materialized_shape[-1] % elements_per_wave != 0
+
+    global_index = remove_thread_indexing(read.index)
+    logger.info(f"global_index={global_index}")
+
+    new_writes = defaultdict(list)
+    for i in range(expected_number_of_loads):
+        # As we adjusted our shape to be in `elements_per_thread` chunks, each
+        # subsequent load will be `total_number_of_threads` elements apart.
+        thread_id_adjusted = thread_id + i * total_number_of_threads
```

Comment:
Sorry Ivan, can you explain this a little more, why is subsequent load `total_number_of_threads` elements apart and not say `total_number_of_threads * old_element_per_thread`?

Reply:
`materialized_shape[-1] // elements_per_thread` made our shape in `elements_per_thread` chunks so all threads will copy `total_number_of_threads` chunks, so the next `GatherToLDS` need to start at `total_number_of_threads` chunk offset.

Reply:
That makes sense, thanks!